File: backend/src/session-store.ts
/**
 * Session Store
 *
 * Manages server-side sessions that outlive WebSocket connections.
 * Each prompt creates a session that buffers all agent events.
 * Sessions persist in memory with a 24h TTL.
 */

import { randomUUID } from 'crypto';
import type { AgentEvent, VaultBridge, SearchResult, FileInfo, GrepResult } from './protocol.js';
import { logger } from './utils.js';

// ============================================================================
// RPC Sender interface — connection-agnostic way to send vault RPCs
// ============================================================================

export interface RpcSender {
  sendRpc<T>(method: string, params: Record<string, unknown>): Promise<T>;
}

// ============================================================================
// Detachable Vault Bridge — can be disconnected/reconnected to different clients
// ============================================================================

export class DetachableVaultBridge implements VaultBridge {
  private sender: RpcSender | null;

  constructor(sender: RpcSender) {
    this.sender = sender;
  }

  get isConnected(): boolean {
    return this.sender !== null;
  }

  detach(): void {
    this.sender = null;
  }

  attach(sender: RpcSender): void {
    this.sender = sender;
  }

  private ensureConnected(): RpcSender {
    if (!this.sender) {
      throw new Error('Client disconnected — vault operations unavailable. Continue with text response only.');
    }
    return this.sender;
  }

  async read(path: string): Promise<string> {
    const sender = this.ensureConnected();
    const result = await sender.sendRpc<{ content: string }>('vault_read', { path });
    return result.content;
  }

  async write(path: string, content: string): Promise<void> {
    const sender = this.ensureConnected();
    await sender.sendRpc('vault_write', { path, content });
  }

  async edit(path: string, oldString: string, newString: string): Promise<void> {
    const sender = this.ensureConnected();
    await sender.sendRpc('vault_edit', { path, old_string: oldString, new_string: newString });
  }

  async search(query: string, limit: number = 20): Promise<SearchResult[]> {
    const sender = this.ensureConnected();
    return sender.sendRpc<SearchResult[]>('vault_search', { query, limit });
  }

  async grep(pattern: string, folder?: string, filePattern?: string, limit: number = 50): Promise<GrepResult[]> {
    const sender = this.ensureConnected();
    return sender.sendRpc<GrepResult[]>('vault_grep', {
      pattern,
      folder: folder || '',
      file_pattern: filePattern || '*.md',
      limit,
    });
  }

  async glob(pattern: string): Promise<string[]> {
    const sender = this.ensureConnected();
    return sender.sendRpc<string[]>('vault_glob', { pattern });
  }

  async list(folder: string): Promise<FileInfo[]> {
    const sender = this.ensureConnected();
    return sender.sendRpc<FileInfo[]>('vault_list', { folder });
  }

  async rename(oldPath: string, newPath: string): Promise<void> {
    const sender = this.ensureConnected();
    await sender.sendRpc('vault_rename', { old_path: oldPath, new_path: newPath });
  }

  async delete(path: string): Promise<void> {
    const sender = this.ensureConnected();
    await sender.sendRpc('vault_delete', { path });
  }
}

// ============================================================================
// Session
// ============================================================================

export type SessionStatus = 'running' | 'complete' | 'error';

type SessionEventCallback = (event: AgentEvent) => void;

export class Session {
  readonly id: string;
  readonly conversationId: string;
  readonly clientId: string;
  readonly prompt: string;
  readonly model: string;
  readonly createdAt: number;

  status: SessionStatus = 'running';
  events: AgentEvent[] = [];
  completedAt?: number;
  error?: string;

  readonly bridge: DetachableVaultBridge;
  private subscribers = new Set<SessionEventCallback>();
  private abortController = new AbortController();

  constructor(params: {
    conversationId: string;
    clientId: string;
    prompt: string;
    model: string;
    sender: RpcSender;
  }) {
    this.id = randomUUID();
    this.conversationId = params.conversationId;
    this.clientId = params.clientId;
    this.prompt = params.prompt;
    this.model = params.model;
    this.createdAt = Date.now();
    this.bridge = new DetachableVaultBridge(params.sender);
  }

  get signal(): AbortSignal {
    return this.abortController.signal;
  }

  /** Push an event — buffers it AND notifies subscribers */
  pushEvent(event: AgentEvent): void {
    this.events.push(event);
    for (const cb of this.subscribers) {
      try {
        cb(event);
      } catch (err) {
        logger.error(`Session ${this.id} subscriber error:`, err);
      }
    }
  }

  /** Subscribe to live events. Returns unsubscribe function. */
  subscribe(callback: SessionEventCallback): () => void {
    this.subscribers.add(callback);
    return () => this.subscribers.delete(callback);
  }

  markComplete(): void {
    this.status = 'complete';
    this.completedAt = Date.now();
    logger.info(`Session ${this.id} completed (${this.events.length} events)`);
  }

  markError(message: string): void {
    this.status = 'error';
    this.completedAt = Date.now();
    this.error = message;
    logger.error(`Session ${this.id} error: ${message}`);
  }

  cancel(): void {
    this.abortController.abort();
  }

  detachBridge(): void {
    this.bridge.detach();
  }

  attachBridge(sender: RpcSender): void {
    this.bridge.attach(sender);
  }

  toInfo(): SessionInfo {
    return {
      sessionId: this.id,
      conversationId: this.conversationId,
      status: this.status,
      createdAt: this.createdAt,
      completedAt: this.completedAt,
      eventCount: this.events.length,
    };
  }
}

export interface SessionInfo {
  sessionId: string;
  conversationId: string;
  status: SessionStatus;
  createdAt: number;
  completedAt?: number;
  eventCount: number;
}

// ============================================================================
// Session Store
// ============================================================================

const SESSION_TTL = 24 * 60 * 60 * 1000; // 24 hours
const CLEANUP_INTERVAL = 60 * 60 * 1000; // 1 hour

export class SessionStore {
  private sessions = new Map<string, Session>();
  private cleanupTimer: ReturnType<typeof setInterval>;

  constructor() {
    this.cleanupTimer = setInterval(() => this.cleanup(), CLEANUP_INTERVAL);
    // Allow the timer to not block process exit
    if (typeof this.cleanupTimer === 'object' && 'unref' in this.cleanupTimer) {
      (this.cleanupTimer as NodeJS.Timeout).unref();
    }
  }

  create(params: {
    conversationId: string;
    clientId: string;
    prompt: string;
    model: string;
    sender: RpcSender;
  }): Session {
    const session = new Session(params);
    this.sessions.set(session.id, session);
    logger.info(`Session ${session.id} created for client ${params.clientId}, conversation ${params.conversationId}`);
    return session;
  }

  get(id: string): Session | null {
    return this.sessions.get(id) || null;
  }

  getByClientId(clientId: string): Session[] {
    return Array.from(this.sessions.values()).filter(s => s.clientId === clientId);
  }

  remove(id: string): void {
    const session = this.sessions.get(id);
    if (session) {
      session.cancel();
      this.sessions.delete(id);
    }
  }

  /** Remove expired sessions */
  cleanup(): void {
    const now = Date.now();
    let removed = 0;
    for (const [id, session] of this.sessions) {
      const age = now - session.createdAt;
      if (age > SESSION_TTL) {
        session.cancel();
        this.sessions.delete(id);
        removed++;
      }
    }
    if (removed > 0) {
      logger.info(`Cleaned up ${removed} expired sessions (${this.sessions.size} remaining)`);
    }
  }

  destroy(): void {
    clearInterval(this.cleanupTimer);
    for (const session of this.sessions.values()) {
      session.cancel();
    }
    this.sessions.clear();
  }

  get size(): number {
    return this.sessions.size;
  }
}

File: backend-py/src/session_store.py
"""Session management — mirrors backend/src/session-store.ts.

Sessions outlive WebSocket connections. Each prompt creates a session
that buffers events. Sessions persist in memory with 24h TTL.
"""

from __future__ import annotations

import asyncio
import time
import uuid
from typing import Any, Callable, Protocol

from .log import log
from .protocol import AgentEvent, event_to_dict

# ============================================================================
# RPC Sender protocol
# ============================================================================

class RpcSender(Protocol):
    async def send_rpc(self, method: str, params: dict[str, Any]) -> Any: ...


# ============================================================================
# Detachable Vault Bridge
# ============================================================================

class DetachableVaultBridge:
    """Vault operations bridge that can be detached/reattached across connections."""

    def __init__(self, sender: RpcSender) -> None:
        self._sender: RpcSender | None = sender

    @property
    def is_connected(self) -> bool:
        return self._sender is not None

    def detach(self) -> None:
        self._sender = None

    def attach(self, sender: RpcSender) -> None:
        self._sender = sender

    def _ensure_connected(self) -> RpcSender:
        if self._sender is None:
            raise RuntimeError(
                "Client disconnected — vault operations unavailable. "
                "Continue with text response only."
            )
        return self._sender

    async def read(self, path: str) -> str:
        result = await self._ensure_connected().send_rpc("vault_read", {"path": path})
        return result.get("content", "") if isinstance(result, dict) else ""

    async def write(self, path: str, content: str) -> None:
        await self._ensure_connected().send_rpc("vault_write", {"path": path, "content": content})

    async def edit(self, path: str, old_string: str, new_string: str) -> None:
        await self._ensure_connected().send_rpc(
            "vault_edit", {"path": path, "old_string": old_string, "new_string": new_string}
        )

    async def search(self, query: str, limit: int = 20) -> list[dict]:
        return await self._ensure_connected().send_rpc(
            "vault_search", {"query": query, "limit": limit}
        )

    async def grep(
        self, pattern: str, folder: str = "", file_pattern: str = "*.md", limit: int = 50
    ) -> list[dict]:
        return await self._ensure_connected().send_rpc(
            "vault_grep",
            {"pattern": pattern, "folder": folder, "file_pattern": file_pattern, "limit": limit},
        )

    async def glob(self, pattern: str) -> list[str]:
        return await self._ensure_connected().send_rpc("vault_glob", {"pattern": pattern})

    async def list(self, folder: str = "") -> list[dict]:
        return await self._ensure_connected().send_rpc("vault_list", {"folder": folder})

    async def rename(self, old_path: str, new_path: str) -> None:
        await self._ensure_connected().send_rpc(
            "vault_rename", {"old_path": old_path, "new_path": new_path}
        )

    async def delete(self, path: str) -> None:
        await self._ensure_connected().send_rpc("vault_delete", {"path": path})


# ============================================================================
# Session
# ============================================================================

EventCallback = Callable[[AgentEvent], None]

SESSION_TTL = 24 * 60 * 60  # 24 hours (seconds)
CLEANUP_INTERVAL = 60 * 60  # 1 hour


class Session:
    def __init__(
        self,
        conversation_id: str,
        client_id: str,
        prompt: str,
        model: str,
        sender: RpcSender,
    ) -> None:
        self.id = str(uuid.uuid4())
        self.conversation_id = conversation_id
        self.client_id = client_id
        self.prompt = prompt
        self.model = model
        self.created_at = int(time.time() * 1000)

        self.status: str = "running"
        self.events: list[dict[str, Any]] = []
        self.completed_at: int | None = None
        self.error_message: str | None = None

        self.bridge = DetachableVaultBridge(sender)
        self._subscribers: set[EventCallback] = set()
        self._cancelled = False

    @property
    def is_cancelled(self) -> bool:
        return self._cancelled

    def push_event(self, event: AgentEvent) -> None:
        event_dict = event_to_dict(event)
        self.events.append(event_dict)
        for cb in list(self._subscribers):
            try:
                cb(event)
            except Exception as e:
                log.error(f"Session {self.id} subscriber error: {e}")

    def subscribe(self, callback: EventCallback) -> Callable[[], None]:
        self._subscribers.add(callback)
        return lambda: self._subscribers.discard(callback)

    def mark_complete(self) -> None:
        self.status = "complete"
        self.completed_at = int(time.time() * 1000)
        log.info(f"Session {self.id} completed ({len(self.events)} events)")

    def mark_error(self, message: str) -> None:
        self.status = "error"
        self.completed_at = int(time.time() * 1000)
        self.error_message = message
        log.error(f"Session {self.id} error: {message}")

    def cancel(self) -> None:
        self._cancelled = True

    def detach_bridge(self) -> None:
        self.bridge.detach()

    def attach_bridge(self, sender: RpcSender) -> None:
        self.bridge.attach(sender)

    def to_info(self) -> dict[str, Any]:
        info: dict[str, Any] = {
            "sessionId": self.id,
            "conversationId": self.conversation_id,
            "status": self.status,
            "createdAt": self.created_at,
            "eventCount": len(self.events),
        }
        if self.completed_at is not None:
            info["completedAt"] = self.completed_at
        return info


# ============================================================================
# Session Store
# ============================================================================

class SessionStore:
    def __init__(self) -> None:
        self._sessions: dict[str, Session] = {}
        self._cleanup_task: asyncio.Task | None = None

    def start_cleanup(self) -> None:
        self._cleanup_task = asyncio.create_task(self._cleanup_loop())

    async def _cleanup_loop(self) -> None:
        while True:
            await asyncio.sleep(CLEANUP_INTERVAL)
            self._cleanup()

    def _cleanup(self) -> None:
        now = int(time.time() * 1000)
        expired = [
            sid for sid, s in self._sessions.items()
            if (now - s.created_at) > SESSION_TTL * 1000
        ]
        for sid in expired:
            s = self._sessions.pop(sid)
            s.cancel()
        if expired:
            log.info(f"Cleaned up {len(expired)} expired sessions ({len(self._sessions)} remaining)")

    def create(
        self,
        conversation_id: str,
        client_id: str,
        prompt: str,
        model: str,
        sender: RpcSender,
    ) -> Session:
        session = Session(conversation_id, client_id, prompt, model, sender)
        self._sessions[session.id] = session
        log.info(
            f"Session {session.id} created for client {client_id}, "
            f"conversation {conversation_id}"
        )
        return session

    def get(self, session_id: str) -> Session | None:
        return self._sessions.get(session_id)

    def get_by_client_id(self, client_id: str) -> list[Session]:
        return [s for s in self._sessions.values() if s.client_id == client_id]

    def destroy(self) -> None:
        if self._cleanup_task:
            self._cleanup_task.cancel()
        for s in self._sessions.values():
            s.cancel()
        self._sessions.clear()

    @property
    def size(self) -> int:
        return len(self._sessions)
